core/services/workflows/syncer/handler.go

```diff
@@ -586,6 +585,226 @@ func (h *eventHandler) tryEngineCleanup(wfID string) error {
+		if err != nil {
+			return fmt.Errorf("failed to get secrets URL hash: %w", err)
+		}
+
+		// Create a new entry in the workflow_spec table corresponding for the new workflow, with the contents of the binaryURL + configURL in the table
+		entry := &job.WorkflowSpec{
+			Workflow:      hex.EncodeToString(decodedBinary),
+			Config:        string(config),
+			WorkflowID:    wfID,
+			Status:        job.WorkflowSpecStatusActive,
+			WorkflowOwner: owner,
+			WorkflowName:  wfMeta.WorkflowName,
+			SpecType:      job.WASMFile,
+			BinaryURL:     wfMeta.BinaryURL,
+			ConfigURL:     wfMeta.ConfigURL,
+		}
+
+		if _, err = h.workflowArtifactsStore.UpsertWorkflowSpecWithSecrets(ctx, entry, wfMeta.SecretsURL, hex.EncodeToString(urlHash), string(secrets)); err != nil {
+			return fmt.Errorf("failed to upsert workflow spec with secrets: %w", err)
+		}
+	}
+
+	// If there is no engine in the engine registry, then create it, start it, and register it
+	engineKey := h.engineRegistry.KeyFor(wfMeta.Owner, wfMeta.WorkflowName)
+	if !h.engineRegistry.Contains(engineKey) {
+		decodedBinary, err := hex.DecodeString(spec.Workflow)
+		if err != nil {
+			return fmt.Errorf("failed to create workflow engine: %w", err)
+		}
+		// Start a new WorkflowEngine instance, and add it to local engine registry
+		engine, err := h.engineFactory(
+			ctx,
+			wfID,
+			owner,
+			workflowName{
+				name: wfMeta.WorkflowName,
+			},
+			[]byte(spec.Config),
+			decodedBinary,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to create workflow engine: %w", err)
+		}
+
+		if err := engine.Start(ctx); err != nil {
+			return fmt.Errorf("failed to start workflow engine: %w", err)
+		}
+
+		if err := h.engineRegistry.Add(engineKey, engine, wfMeta); err != nil {
+			// This shouldn't happen because we call the handler serially and
+			// check for running engines above, see the call to engineRegistry.Contains.
+			return fmt.Errorf("invariant violation: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// removeWorkflow handles deleting a workflow from contract workflow metadata. This method must remain idempotent.
+func (h *eventHandler) removeWorkflow(
+	ctx context.Context,
+	wfMeta GetWorkflowMetadata,
+) error {
+	var multiErr error
+
+	if err := h.tryEngineCleanup(wfMeta.Owner, wfMeta.WorkflowName); err != nil {
+		multiErr = multierr.Append(multiErr, err)
+	}
+
+	if err := h.workflowArtifactsStore.DeleteWorkflowArtifacts(ctx, hex.EncodeToString(wfMeta.Owner),
+		wfMeta.WorkflowName, hex.EncodeToString(wfMeta.WorkflowID[:])); err != nil {
+		multiErr = multierr.Append(multiErr, fmt.Errorf("failed to delete workflow artifacts: %w", err))
+	}
+
+	return multiErr
+}
+
+// HandleWorkflowMetadata compares the latest contract state for workflow metadata
+// against what is currently running on the node via the engine registry
+func (h *eventHandler) HandleWorkflowMetadata(ctx context.Context, state []GetWorkflowMetadata) {
```

Comment:
@justinkaseman Hmmm... is this necessary? What data does state give us that we won't get via the event struct?

With the addition of the feature flag we want to keep all of the code that deals with the two modes as close and concentrated as possible, otherwise the code will be incredibly hard to reason about.

Reply:
Cleaned up. Check it out again.